---
title: "LAO_tree"
author: "Francesco Delogu"
date: "10/8/2021"
output: html_document
---
# Analysis of LAO phylogenetic tree

## Set the environment

```{r setup}

options(stringsAsFactors = F, gsubfn.engine = "R")

library(tidyverse)
library(ggtree)
library(treeio)
library(ape)
library(phytools)
library(castor)
library(ggridges)
library(ggtreeExtra)
library(ggnewscale)
library(viridis)
library(ggbreak)

wd <- getwd()

```

## Load data

```{r Load data, Echo=FALSE, cache=FALSE, results=FALSE, warning=FALSE, comment=FALSE, warning=FALSE}

bac.gtdbtk <- read.newick(paste0(wd, "/data/gtdbtk.bac120.classify.tree"))
MAGs.checkm <- read.csv(paste0(wd, "/data/derepgenomeInfo.csv"))

bin.tax <- read_tsv(paste0(wd, "/data/out.BAT.classification")) %>%
  `colnames<-`(c("bin", colnames(.)[-1])) %>%
  separate(superkingdom, c("superkingdom", "to_delete"), sep=":") %>%
  separate(phylum, c("phylum", "to_delete"), sep=":") %>%
  separate(class, c("class", "to_delete"), sep=":") %>%
  separate(order, c("order", "to_delete"), sep=":") %>%
  separate(family, c("family", "to_delete"), sep=":") %>%
  separate(genus, c("genus", "to_delete"), sep=":") %>%
  separate(species, c("species", "to_delete"), sep=":") %>%
  dplyr::select(-to_delete, -lineage, -`lineage scores`, -reason, -classification)

bin.map <- read_tsv(paste0(wd, "/data/bin_summary.txt"), col_names=F) %>%
  `colnames<-`(c("contig", "bin"))

contig.tax <- read_tsv(paste0(wd, "/data/out.CAT2.classification")) %>%
  `colnames<-`(c("contig", colnames(.)[-1])) %>%
  separate(superkingdom, c("superkingdom", "to_delete"), sep=":") %>%
  separate(phylum, c("phylum", "to_delete"), sep=":") %>%
  separate(class, c("class", "to_delete"), sep=":") %>%
  separate(order, c("order", "to_delete"), sep=":") %>%
  separate(family, c("family", "to_delete"), sep=":") %>%
  separate(genus, c("genus", "to_delete"), sep=":") %>%
  separate(species, c("species", "to_delete"), sep=":") %>%
  dplyr::select(-to_delete, -lineage, -`lineage scores`, -reason) %>%
  mutate(to_split=contig) %>%
  separate(to_split, c("to_delete", "to_split"), sep="length_") %>%
  separate(to_split, c("Length", "to_delete"), sep="_cov") %>%
  dplyr::select(-c("to_delete"))

contig.bin.tax <- full_join(bin.map, bin.tax) %>%
  separate(contig, c("to_delete", "to_split"), sep="length_", remove=F) %>%
  separate(to_split, c("Length", "to_delete"), sep="_cov") %>%
  dplyr::select(-c("to_delete"))

contig.cov <- read_tsv(paste0(wd, "/data/contig_cov.tsv")) %>%
  mutate(contig=...1) %>%
  dplyr::select(-...1)

cov.summ <- (inner_join(as.data.frame(rbind(contig.bin.tax %>%
                                  dplyr::select(contig, bin, phylum),
                                contig.tax %>%
                                  mutate(bin="Unbinned") %>%
                                  dplyr::select(contig, bin, phylum))),
                      contig.cov) %>%
  mutate(bin=substr(bin, 1, str_length(bin)-3)) %>%
  pivot_longer(names_to="Sample", values_to="Coverage", -c(contig, bin, phylum)) %>%
  filter(startsWith(Sample, "D")) %>%
  group_by(contig, bin, phylum) %>%
  dplyr::summarise(Coverage=mean(Coverage), .groups="keep") %>%
  ungroup())[,c("bin", "contig", "phylum", "Coverage")]

LAO_samples <- read.table(paste0(wd, "/data/samples.txt"), sep=" ", header=T)
sample_to_date <- as.Date(LAO_samples$Date)
names(sample_to_date) <- LAO_samples$Sample

```

## Plot tree

```{r Plot tree}

sub.bac.gtdbtk <- (bac.gtdbtk %>%
  get_subtree_with_tips(only_tips=substr(MAGs.checkm$genome, 1, str_length(MAGs.checkm$genome)-3), collapse_monofurcations=T))$subtree

sub.bac.gtdbtk$node.label <- 1:length(sub.bac.gtdbtk$node.label)

p <- ggtree(sub.bac.gtdbtk, layout="circular") %<+%
  (MAGs.checkm %>%
     mutate(genome=substr(genome, 1, str_length(genome)-3))) %<+%
  (bin.tax %>%
     mutate(bin=substr(bin, 1, str_length(bin)-3)))

p.plot <- p +
  #geom_tippoint(aes(size=completeness, color=phylum), shape=1) +
  geom_tiplab() +
  geom_nodelab()
  
ggsave(paste0(wd, "/tree_node_labels.png"), p.plot, dpi=320, height=20, width=20)

p +
  geom_tippoint(aes(size=completeness), shape=1) +
  geom_strip("D14__O1.25_sub.contigs", "D03__P2.50.2.contigs", color="gold", barsize=2, extend=0.2) + # Actinobacteria
  geom_strip("D20__17813.contigs", "D20__17813.contigs", color="chartreuse4", barsize=2, extend=0.2) + # Proteobacteria
  geom_strip("F13__169.contigs", "D02__7837.contigs", color="cornflowerblue", barsize=2, extend=0.2) + # Spirochetes
  geom_strip("D08__G3.contigs", "D36__G15.contigs", color="darkred", barsize=2, extend=0.2) + # Bacteroidetes
  geom_strip("D27__1477.contigs", "D27__1477.contigs", color="blue", barsize=2, extend=0.2) + # Nitrospirae
  geom_strip("D30__9407_sub.contigs", "F13__G3.1.2.70.contigs", color="darkgreen", barsize=2, extend=0.2) + # Proteobacteria
  geom_strip("D38__O10.contigs", "D38__O10.contigs", color="black", barsize=2, extend=0.2) + # Candidatus Gracilibacteria
  geom_strip("F14__P3.2.contigs", "F14__P3.2.contigs", color="darkorchid", barsize=2, extend=0.2) + # Fusobacteria
  geom_strip("D51__P2.16.contigs", "D51__P2.16.contigs", color="orange", barsize=2, extend=0.2) # Chloroflexi

```


## Plot cov

```{r Plot cov}

library(wesanderson)

selected.phyla <- (cov.summ %>%
  group_by(bin, phylum) %>%
  dplyr::summarise(value=max(Coverage)) %>%
  ungroup() %>%
  filter(value > 3))$phylum

short_phyla <- c("Acidobacteria", "Actinobacteria", "Bacteroidetes", "Ca. Eisenbacteria", "Ca. Gracilibacteria", "Ca. Moranbacteria", "Ca. Saccharibacteria", "Chlorobi", "Chloroflexi", "Ciliophora", "Cyanobacteria", "Euryarchaeota", "Evosea", "Firmicutes", "Fusobacteria", "Gemmatimonadetes", "Ignavibacteriae", "Nitrospirae", "Planctomycetes", "Proteobacteria", "Rotifera", "Spirochaetes", "Streptophyta", "Verrucomicrobia")
names(short_phyla) <- c("Acidobacteria", "Actinobacteria", "Bacteroidetes", "Candidatus Eisenbacteria", "Candidatus Gracilibacteria", "Candidatus Moranbacteria", "Candidatus Saccharibacteria", "Chlorobi", "Chloroflexi", "Ciliophora", "Cyanobacteria", "Euryarchaeota", "Evosea", "Firmicutes", "Fusobacteria", "Gemmatimonadetes", "Ignavibacteriae", "Nitrospirae", "Planctomycetes", "Proteobacteria", "Rotifera", "Spirochaetes", "Streptophyta", "Verrucomicrobia")

pal_binned <- wes_palette("Cavalcanti1", 2)

binunbin.plot <- rbind((contig.tax %>%
         mutate(bin="Unbinned") %>%
         dplyr::select(phylum, bin, Length)),
      (contig.bin.tax %>%
         mutate(bin="Binned") %>%
         dplyr::select(phylum, bin, Length))) %>%
  group_by(phylum, bin) %>%
  summarise(Sum_Length=sum(as.numeric(Length)), .groups="keep") %>%
  filter(Sum_Length>1e+06, !phylum%in%c(NA, "not classified")) %>%
  pivot_wider(names_from=bin, id_cols=phylum, values_from=Sum_Length) %>%
  mutate(Unbinned=ifelse(is.na(Unbinned), 0, Unbinned), Binned=ifelse(is.na(Binned), 0, Binned), phylum=short_phyla[phylum]) %>%
  filter(Unbinned>1e+06 | Binned>1e+06) %>%
  pivot_longer(names_to="bin", values_to="Sum_Length", -phylum) %>%
  ggplot(aes(x=factor(phylum, levels=sort(unique(phylum), decreasing=F)), y=Sum_Length+1, fill=bin)) +
  geom_bar(stat="identity", position="dodge2") +
  scale_fill_manual(values=pal_binned) +
  scale_y_log10(breaks=c(0, 1e+01)) +
  scale_y_break(c(1.1e+02, 1.0e+05), scales=9, ticklabels=c(1.0e+05, 1.0e+06, 1.0e+07, 1.0e+08)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 60, hjust = 1),
        legend.position="top",
        panel.grid.major.y = element_line(color = "grey80")) +
  labs(x="Phylum (>1e+06 Nt)", y="Total MG sequence length [Nt]")

binunbin.plot

ggsave(paste0(wd, "/binunbin.svg"), binunbin.plot, dpi=320, height=4, width=6)
ggsave(paste0(wd, "/binunbin.png"), binunbin.plot, dpi=320, height=4, width=6)

``` 


## Plot cov

```{r Plot cov}

p <- ggtree(sub.bac.gtdbtk, layout = "fan", open.angle=25, size=0.5)
p <- rotate_tree(p, +90)
p <- p %<+%
  (bin.tax %>%
     mutate(bin=substr(bin, 1, str_length(bin)-3)))


cov.summ2 <- cov.summ %>%
  mutate(Phylum=phylum) %>%
  filter(Coverage > 0.05 & Coverage < 2)
MAGs.checkm2 <- MAGs.checkm %>%
  mutate(genome=substr(genome, 1, str_length(genome)-3)) %>%
  dplyr::select(-c(length, N50)) %>%
  pivot_longer(names_to="bin.vars", values_to="value", -genome)

pal_phyla <- wes_palette("Darjeeling1", 9, type="continuous")

tree.plot <- p +
  geom_tiplab(align=TRUE, linesize=.25, size=0) +
  geom_tippoint(shape=21, size=2.5,
                mapping=aes(fill=phylum)) +
  scale_fill_manual(values=pal_phyla) +
  new_scale_fill() +
  #geom_nodelab() +
  #geom_highlight(node=c("28"), fill="darkgreen", alpha=0.5) +
  geom_strip("D26__maxbin_res.001.fasta.contigs", "D02__15476.contigs",
             color = "black", barsize = 1.5, extend = 0.2,
             label="*   ", fontsize = 6, offset.text = 0.05, angle = -45) + # Candidatus Microthrix
  geom_strip("D30__9407_sub.contigs", "D18__maxbin_res.002.fasta.contigs",
             color = "black", barsize = 1.5, extend = 0.2,
             label="**", fontsize = 6, offset.text = 0.02, angle = -45) + # Moraxellaceae bacterium AER2_44_116
  geom_fruit(data = MAGs.checkm2,
             geom = geom_tile,
             aes(y=genome, x=bin.vars, fill=value),
             color = "white", size = 2,
             grid.params = list(),
             axis.params = list(axis="x", text.size = 4, hjust=-0.05),
             offset = 0.05,
             pwidth = 0.1) +
  scale_fill_viridis()  +
  new_scale_fill() +
  geom_fruit(data = cov.summ2,
             geom = geom_violin,
             aes(x=Coverage, group=label, y=bin, fill=Phylum),
             size=0.5,
             grid.params=list(),
             axis.params=list(axis="x", text.size= 4, hjust=-0.1, vjust=0.5, nbreak=3,),
             offset = 0.1,
             pwidth = 0.2) +
  scale_fill_manual(values=pal_phyla) +
  theme_tree2() +
  theme(legend.position = 0)

ggsave(paste0(wd, "/tree.svg"), tree.plot, dpi=320, height=10, width=10)
ggsave(paste0(wd, "/tree.png"), tree.plot, dpi=320, height=15, width=15)


``` 

